src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java

```diff
@@ -386,20 +419,45 @@ protected long findLatestFileNameGeneration(ImmutableMap<String, BlobMetaData> b
          * @param blobs list of blobs in repository
          * @return BlobStoreIndexShardSnapshots
          */
-        protected BlobStoreIndexShardSnapshots buildBlobStoreIndexShardSnapshots(ImmutableMap<String, BlobMetaData> blobs) {
-            List<BlobStoreIndexShardSnapshot> snapshots = Lists.newArrayList();
+        protected Tuple<BlobStoreIndexShardSnapshots, Integer> buildBlobStoreIndexShardSnapshots(ImmutableMap<String, BlobMetaData> blobs) {
+            int latest = -1;
+            for (String name : blobs.keySet()) {
+                if (name.startsWith(SNAPSHOT_INDEX_PREFIX)) {
+                    try {
+                        int gen = Integer.parseInt(name.substring(SNAPSHOT_INDEX_PREFIX.length()));
+                        if (gen > latest) {
+                            latest = gen;
+                        }
+                    } catch (NumberFormatException ex) {
+                        logger.warn("failed to parse index file name [{}]", name);
+                    }
+                }
+            }
+            if (latest >= 0) {
+                try (InputStream stream = blobContainer.openInput(SNAPSHOT_INDEX_PREFIX + latest)) {
+                    try (XContentParser parser = XContentFactory.xContent(XContentType.JSON).createParser(stream)) {
+                        parser.nextToken();
+                        return new Tuple<>(BlobStoreIndexShardSnapshots.fromXContent(parser), latest);
```

Comment:
I cannot reuse it here because this method reads multiple snapshots. But I can definitely add a new method `readSnapshots()` similiar to `readSnapshot()` and use it here.